### PR TITLE
improve efficiency of reflector

### PR DESCRIPTION
**Pull request by juliantaylor**
The reflector performs a _list_and_update every watch timeout (default 10s) requesting most recent data from the apiserver via the default resource_version="".

This requires the apiserver to retrieve the latest data from the backing store (etcd) which can be very expensive in large clusters. For example if omit_namespace is set the full cluster pod listings require several seconds retrieve all data from the etcd in clusters with more then 10000 pods. The label selector filtering is only done in the apiserver.

As the _list_and_update is used for watches the listing should set the resource version and resource version match to tell the apiserver that its locally cached data is sufficient.
The initial listing will fetch any data the apiserver currently has to prepopulate the watch, the watch itself will then fetch whatever is remaining from the etcd.

The subsequent relistings after the watch timeouts use their last received resource version in the same way.

For details see https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions

Closes gh-753

### CODE REVIEW

File: kubespawner/reflector.py [hunk 1]
```diff
@@ -208,7 +208,7 @@ def __init__(self, *args, **kwargs):
 
         self.watch_task = None
 
-    async def _list_and_update(self):
+    async def _list_and_update(self, resource_version=""):
```
Comment (yuvipanda) [inline kubespawner/reflector.py:211]:
```suggestion
    async def _list_and_update(self, resource_version=None):
```

Reading through https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions, I see that there's a difference between this being unset, and being set to the string "0" but not for the empty string "". 

So I think we should either use `None` here to indicate that it will be *unset* if not specified, or "0" to get the behavior for that. As "" doesn't seem to be defined, I don't think we should default to that.

I'm not super set on `None` or `"0"`, we can go with whatever you think works better here :)

File: kubespawner/reflector.py [hunk 2]
```diff
@@ -218,9 +218,12 @@ async def _list_and_update(self):
         kwargs = dict(
             label_selector=self.label_selector,
             field_selector=self.field_selector,
+            resource_version=resource_version,
             _request_timeout=self.request_timeout,
             _preload_content=False,
         )
+        if resource_version:
```
Comment (yuvipanda) [inline kubespawner/reflector.py:225]:
```suggestion
        if resource_version is not None:
```

File: kubespawner/reflector.py [hunk 3]
```diff
@@ -218,9 +218,12 @@ async def _list_and_update(self):
         kwargs = dict(
             label_selector=self.label_selector,
             field_selector=self.field_selector,
+            resource_version=resource_version,
```
Comment (yuvipanda) [inline kubespawner/reflector.py:221]: Should we set this only if it's not `None`?

File: kubespawner/reflector.py [hunk 4]
```diff
@@ -264,6 +267,9 @@ async def _watch_and_update(self):
             selectors.append("field selector=%r" % self.field_selector)
         log_selector = ', '.join(selectors)
 
+        # fetch Any (=api-server cached) data from apiserver on initial fetch
```
Comment (yuvipanda) [inline kubespawner/reflector.py:270]:
```suggestion
        # fetch Any (=api-server cached) data from apiserver on initial fetch
        # see https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
        # for more information
```

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Pretty excited for this PR, @juliantaylor! Thank you for taking the time, I appreciate it :)

I've left one review comment about the function default, but otherwise, given the tests pass, I am pretty happy with this. If you can resolve that comment one way or another, unless someone else objects, I'll be happy to merge this in a couple days.

File: kubespawner/reflector.py (hunk 1 above)
Comment (juliantaylor) [reply kubespawner/reflector.py:211]: it is not really clear what unset means, in the code resource version seems to always just be a string and it is compared against an empty string which I interpret as meaning unset.

then another issue is what the python api library does with None vs a string but I assume it is likely equivalent so using None may be clearer

File: kubespawner/reflector.py (hunk 2 above)
Comment (juliantaylor) [reply kubespawner/reflector.py:225]: this is intentional so that it also does not match the empty string which I'm pretty sure also means unset

File: kubespawner/reflector.py (hunk 3 above)
Comment (juliantaylor) [reply kubespawner/reflector.py:221]: As it is confusing what unset means in the api and its clients Ill update the code to use None and explicitly not add it to the kwargs

File: kubespawner/reflector.py [hunk 5]
```diff
@@ -218,9 +218,12 @@ async def _list_and_update(self):
         kwargs = dict(
             label_selector=self.label_selector,
             field_selector=self.field_selector,
+            resource_version=resource_version,
             _request_timeout=self.request_timeout,
```
Comment (juliantaylor) [inline kubespawner/reflector.py:222]: I do think it would be a good idea to change this timeout to a larger value for the watches, even if cached lists are still very expensive. Other k8s components typically relist watches in the range of minutes to never, every 10 seconds is a bit much.
Though probably should be done in a different PR.

File: kubespawner/reflector.py (hunk 1 above)
Comment (yuvipanda) [reply kubespawner/reflector.py:211]: In the page you linked to, I see a column in the table for `resourceVersion not set`. I am slightly confused about what you mean by "it is not really clear what unset means". By "in the code", do you mean in this codebase or in kubernetes?

Another alternative is to just require `resource_version` to always be passed in here, and not let that be optional.

File: kubespawner/reflector.py (hunk 5 above)
Comment (yuvipanda) [reply kubespawner/reflector.py:222]: Yep let's consider that in a different PR

**Review by minrk — APPROVED** [review]

File: kubespawner/reflector.py (hunk 1 above)
Comment (juliantaylor) [reply kubespawner/reflector.py:211]: yes I meant the kubernetes code, the resource version there is a string which either has an opaque value (which happens to be etcd's modification index) or is empty

empty means unset

File: kubespawner/reflector.py [hunk 6]
```diff
@@ -346,6 +355,9 @@ async def _watch_and_update(self):
                 self.log.debug("Cancelled watching %s", self.kind)
                 raise
             except Exception:
+                # ensure we request a valid resource version on retry,
+                # needed on 410 Gone errors
+                resource_version = "0"
```
Comment (juliantaylor) [inline kubespawner/reflector.py:360]: added this, should not be needed with the `NotOlderThan` resource_version_match but it probably makes sense to start from the beginning on errors.

**Comment by yuvipanda** [comment]
Thanks, @juliantaylor!
